homeassistant/components/sun.py

```diff
@@ -51,172 +51,143 @@ def is_on(hass, entity_id=None):
     return hass.states.is_state(entity_id, STATE_ABOVE_HORIZON)
 
 
-def next_dawn(hass, entity_id=None):
+def next_dawn(hass, entity_id=None, offset=None):
     """Local datetime object of the next dawn.
 
     Async friendly.
     """
-    utc_next = next_dawn_utc(hass, entity_id)
+    utc_next = next_dawn_utc(hass, entity_id, offset)
 
     return dt_util.as_local(utc_next) if utc_next else None
 
 
-def next_dawn_utc(hass, entity_id=None):
+def next_dawn_utc(hass, entity_id=None, offset=None):
     """UTC datetime object of the next dawn.
 
     Async friendly.
     """
-    entity_id = entity_id or ENTITY_ID
-
-    state = hass.states.get(ENTITY_ID)
-
-    try:
-        return dt_util.parse_datetime(
-            state.attributes[STATE_ATTR_NEXT_DAWN])
-    except (AttributeError, KeyError):
-        # AttributeError if state is None
-        # KeyError if STATE_ATTR_NEXT_DAWN does not exist
+    entity = hass.data.get(entity_id or ENTITY_ID)
+    if entity is None:
         return None
 
+    return Sun.get_next_solar_event(
```

Comment:
I've tried to lay out a few points of why I think the sun component is fundamentally different from any of the other components. Don't want to prolong this thread, so I've tried to collect my thoughts here. Sorry for the big dump, but this will be the last I have to say on it.

1) The astral component isn't talking to any devices, it isn't performing IO, it's just doing math on its inputs.
2) The static inputs to astral are latitude, longitude, and elevation, all available from the hass config.
3) The only input to astral that changes is the date. For any number of remote instances, if any of them provide these sames inputs to astral they will get the same results. There is no actual state that needs to be stored. Since we already allow all instances to use system time from the python `datetime` module, each instance already has all the information it needs to calculate these events.
4) The biggest giveaway that the sun is different, is we allow offsets to events. We obviously don't allow triggers 1 hour before an MQTT event. If this was a normal component, we'd have to wait for the sun rising event on the event bus like normal. However, we have this functionality because sun events aren't an event that we're waiting for to happen, when it will happen is just a fact that anyone can calculate and get the same result.

I built this implementation assuming we want to make astral a core dependency. It isn't necessarily a requirement, but given the number of interconnections that exist, it might be worth considering. Alternatively, we can just require any component that calls the sun helper methods depend on astral.

Also worth noting: As part of this rework I discovered that the flux component was making the same error as the automation condition logic.